include/small/util.h

```diff
@@ -58,3 +113,271 @@ small_getpagesize(void)
+	wrapper->payload = small_wrapper_payload(wrapper->ptr, header_size,
+						 payload_size, alignment);
+	wrapper->header = small_wrapper_header(wrapper->payload, header_size);
+	small_assert((char *)wrapper->header >= wrapper->ptr);
+	wrapper->header->alloc_offset = (char *)wrapper->header - wrapper->ptr;
+	/*
+	 * This poison is expected to be precise (without alignment issues)
+	 * because we poison to the end of the allocated block.
+	 */
+	char *begin = wrapper->payload + payload_size;
+	char *end = wrapper->ptr + wrapper_size;
+	ASAN_POISON_MEMORY_REGION(begin, end - begin);
+}
+
+/**
+ * Unpoison wrapper header memory. Unpoison is intended to be used
+ * on freeing memory so we don't need to unpoison what we poisoned
+ * before entirely. We need to unpoison only header to be able to
+ * read header data to do extra actions before freeing memory.
+ */
+static inline void
+small_wrapper_unpoison(char *payload, void *header)
+{
+	/* Check unpoison actually unpoison from the beginning of the header. */
+	static_assert(SMALL_HEADER_ALIGNMENT % SMALL_POISON_ALIGNMENT == 0,
+		      "header should be aligned on ASAN alignment");
+	/* Unpoison both header and magic area if the latter exists. */
+	ASAN_UNPOISON_MEMORY_REGION(header, payload - (char *)header);
+	char *magic_begin = (char *)small_align_down((uintptr_t)payload,
+						     SMALL_POISON_ALIGNMENT);
+	small_assert(memcmp(magic_begin, &small_magic,
+		     payload - magic_begin) == 0 && "wrapper magic check");
+}
+
+/**
+ * Free wrapper memory.
+ */
+static inline void
+small_wrapper_free(struct small_wrapper *wrapper)
+{
+	/* Call unpoison to check that magic is not changed. */
+	small_wrapper_unpoison(wrapper->payload, wrapper->header);
+	free(wrapper->ptr);
+}
+
+/**
+ * Initialize wrapper pointers given header pointer. Alignment argument
+ * is payload alignment.
+ */
+__attribute__((no_sanitize_address))
+static inline void
+small_wrapper_from_header(struct small_wrapper *wrapper, void *header,
+			  size_t payload_size, size_t alignment,
+			  size_t header_size)
+{
+	wrapper->header = (struct small_header *)header;
+	wrapper->ptr = (char *)header - wrapper->header->alloc_offset;
+	wrapper->payload =
+		small_wrapper_payload(wrapper->ptr, header_size, payload_size,
+				      alignment);
+}
+
+/**
+ * Initialize wrapper pointers given payload pointer. Alignment argument
+ * is payload alignment. Additionally header is unpoisoned.
+ *
+ * Intended usage on freeing memory is:
+ *    small_wrapper_from_payload(&wrapper, payload, ..);
+ *    Do some actions accessing wrapper.header.
+ *    small_wrapper_free(&wrapper);
+ */
+__attribute__((no_sanitize_address))
+static inline void
+small_wrapper_from_payload(struct small_wrapper *wrapper, void *payload,
+			   size_t header_size)
+{
+	wrapper->payload = (char *)payload;
+	wrapper->header = small_wrapper_header(wrapper->payload, header_size);
+	small_wrapper_unpoison(wrapper->payload, wrapper->header);
```

Comment:
The code-flow paths look asymmetrical, which is confusing:

- `small_wrapper_from_payload` calls `small_wrapper_unpoison` while `small_wrapper_from_header` doesn't.
- `small_wrapper_alloc` doesn't call `small_wrapper_poison` while `small_wrapper_free` does call `small_wrapper_unpoison`.
- `small_wrapper_free` is sometimes called after `small_wrapper_from_payload`, calling `small_wrapper_unpoison` for the second time for no reason.

Reply:
API for allocation memory is changed as described in one of the previous comments.